pyccel/ast/internals.py

```diff
@@ -163,3 +162,74 @@ def __str__(self):
             stop = str(self.stop)
         return '{0} : {1}'.format(start, stop)
 
+class Symbol(Basic):
+    """
+    Represent variable with undefined type
+
+    Parameters
+    ----------
+    name : String
+        name of the symbol
+
+    Examples
+    --------
+    >>> from pyccel.ast.internals import Symbol
+    >>> x = Symbol('x')
+    x
+    """
+    def __init__(self, name):
+        if not isinstance(name, str):
+            raise TypeError('Symbol name should be a string, not '+ str(type(name)))
+        self._name = name
+        super().__init__()
+
+    @property
+    def name(self):
+        """Name of the symbol
+        """
+        return self._name
+
+    def __eq__(self, other):
+        if isinstance(other, str):
+            return self._name == other
+        elif type(self) is type(other):
+            return self._name == other.name
+        else:
+            return False
+
+    def __hash__(self):
+        return hash(self._name)
```

Comment:
If 2 objects with the same name but different classes are not equal then shouldn't they have different hashes?

Reply:
seem logic but i wanted to make a string hash the same as the symbol one

Reply:
Any reason for that?

Normally if hashes are the same, the objects are equal. I looked into it briefly and found [this link](https://hynek.me/articles/hashes-and-equality/) which says:

> Hashes can be less picky than equality checks. Since key lookups are always followed by an equality check, your hashes don’t have to be unique. That means that you can compute your hash over an immutable subset of attributes that may or may not be a unique “primary key” for the instance.
> 
> You can take this property to town by returning a constant hash and make the set or dictionary work purely on equality checks. However that would regress them into a list and lead to terrible performance.

So apparently it's ok but we are likely to lose performance which isn't great.

If the pyccel `Symbol` and the sympy `Symbol` need to have the same hash for some reason, then I am inclined to agree with Yaman that `Variable` shouldn't inherit from `Symbol` so different `Variable` classes can have different hashes